pkg/codegen/importer/hcl2.go

```diff
@@ -425,7 +426,68 @@ func generateValue(typ schema.Type, value resource.PropertyValue) (model.Express
 					})
 				}
 			}
-		} else {
+		case *schema.UnionType:
+			discriminatorPropertyName := objectType.Discriminator
+			selectElementType := func() (*schema.ObjectType, bool) {
+				discriminatorValue, ok := obj[resource.PropertyKey(discriminatorPropertyName)]
+				if !ok {
+					// discriminator property is not present, so we can't select a type
+					return nil, false
+				}
+
+				if !discriminatorValue.IsString() {
+					// discriminator property value is not a string,
+					// so we can't select a type from the union mapping
+					return nil, false
+				}
+
+				discriminatorValueText := discriminatorValue.StringValue()
+				correspondingTypeToken, ok := objectType.Mapping[discriminatorValueText]
+				if ok {
+					for _, elementType := range objectType.ElementTypes {
+						// found the type token
+						// match it against the element type which should be an object
+						elementTypeObject, ok := codegen.UnwrapType(elementType).(*schema.ObjectType)
+						if ok {
+							elementTypeParts := strings.Split(elementTypeObject.Token, ":")
```

Comment:
Can we replace this by casting to [`tokens.Type`](https://github.com/pulumi/pulumi/blob/adef574983fa6f4e1ab48e50ef5d1388da1efcdb/sdk/go/common/tokens/tokens.go#L212) and then calling the appropriate methods?